File: frontend/src/services/auth.service.ts
import api from "./api";
import type { User, TokenResponse } from "../types";

export const TOKEN_KEY = "access_token";
export const REFRESH_TOKEN_KEY = "refresh_token";

export const register = async (username: string, email: string, password: string): Promise<User> => {
  const { data } = await api.post<User>("/auth/register", { username, email, password });
  return data;
};

export const login = async (email: string, password: string): Promise<User> => {
  const { data } = await api.post<TokenResponse>("/auth/login", { email, password });
  localStorage.setItem(TOKEN_KEY, data.access_token);
  localStorage.setItem(REFRESH_TOKEN_KEY, data.refresh_token);
  const { data: user } = await api.get<User>("/auth/me");
  return user;
};

export const logout = async (): Promise<void> => {
  try {
    await api.post("/auth/logout");
  } finally {
    localStorage.removeItem(TOKEN_KEY);
    localStorage.removeItem(REFRESH_TOKEN_KEY);
  }
};

export const getMe = async (): Promise<User> => {
  const { data } = await api.get<User>("/auth/me");
  return data;
};

File: backend/app/services/auth_service.py
from fastapi import HTTPException, status

from app.core.security import hash_password, verify_password, create_access_token, create_refresh_token
from app.db.repositories.user_repository import UserRepository
from app.models.user import UserCreate, UserLogin, TokenResponse


class AuthService:
    def __init__(self, repo: UserRepository):
        self.repo = repo

    async def register(self, user_data: UserCreate) -> dict:
        if await self.repo.get_by_email(user_data.email):
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Email already registered")
        if await self.repo.get_by_username(user_data.username):
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Username already taken")
        data = user_data.model_dump()
        data["hashed_password"] = hash_password(data.pop("password"))
        data["email"] = data["email"].lower()
        data["is_active"] = True
        return await self.repo.create(data)

    async def login(self, credentials: UserLogin) -> TokenResponse:
        user = await self.repo.get_by_email(credentials.email)
        if not user or not verify_password(credentials.password, user["hashed_password"]):
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid email or password")
        token_data = {"sub": user["id"]}
        return TokenResponse(
            access_token=create_access_token(token_data),
            refresh_token=create_refresh_token(token_data),
        )
